freedomofpress/securedrop
# Uses `max_fail_percentage` to force immediate Ansible exits

## Status

Ready for review.

## Description of Changes

Fixes #2885.

Changes proposed in this pull request:

* sets `max_fail_percentage=0` on each and every play in all Ansible playbooks
* includes regressions tests to make sure any new plays include `max_fail_percentage=0`
* removes unused, nearly-three-year-old "upgrade" logic designed for SecureDrop 0.3
* uses new "ansible-config" Molecule scenario appropriate for local env testing (can be expanded in the future)

## Testing

There are actually two things we want to test:

1. Errors on one host during a prod run should cause immediate exit.
2. Any changes to Ansible playbooks that omit `max_fail_percentage: 0` should be caught by CI and cause the test suite to fail.

The first is the important one. It's not possible to test with staging, since staging already assumes that both hosts are present (via the ci-tweaks role) and will indeed fail fast if not. So, use prod VMs! Then:

1. Run `./securedrop-admin install`
2. During execution of the "common" role, perform hard shutdown on `mon-prod`.
3. Confirm that the entire playbook exits immediately, _without_ continuing to execute against `app-prod`.

Additional output for comparing the behavior under the develop branch can be found in https://github.com/freedomofpress/securedrop/issues/2885#issuecomment-360275071.

## Deployment

Designed to implement a "fail fast" strategy for production installs. Any error should cause immediate exit of the configuration run.

## Checklist

### If you made changes to the app code:


### If you made changes to the system configuration:


### If you made changes to documentation:
